rich/progress.py

```diff
@@ -432,34 +431,27 @@ def start(self) -> None:
 
     def stop(self) -> None:
         """Stop the progress display."""
+        assert self._refresh_thread is not None
         with self._lock:
             if not self._started:
                 return
             self._started = False
             try:
-                if self.auto_refresh and self._refresh_thread is not None:
+                if self.auto_refresh:
                     self._refresh_thread.stop()
-                    self._refresh_thread = None
                 self.refresh()
                 if self.console.is_terminal:
                     self.console.line()
             finally:
                 self.console.show_cursor(True)
+        self._refresh_thread.join()
 
     def __enter__(self) -> "Progress":
-        with self._lock:
-            if self._enter_count:
-                self._enter_count += 1
-                return self
-            self.start()
-            self._enter_count += 1
-            return self
+        self.start()
```

Comment:
Pretty sure there is no practical use for a re-entrant context manager here. Just complicated things.